From 313900c37ce5255ef6aad736d6776f8e89076c59 Mon Sep 17 00:00:00 2001
From: auto commit <unknown@unknown>
Date: Fri, 24 Oct 2008 12:22:35 -0700
Subject: [PATCH] cavium: timer IRQ cleanup

Kernel.org git ID 7bcf7717 for clockevents totally changes the timer
code.  This brings it in line with the other r4k read_c0 type timer
platforms and reduces the board specifics even more.

Signed-off-by: Paul Gortmaker <Paul.Gortmaker@windriver.com>
---
 arch/mips/cavium-octeon/setup.c |   69 ++------------------------------------
 1 files changed, 4 insertions(+), 65 deletions(-)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index fdb6898..61af4db 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -4,6 +4,7 @@
  * for more details.
  *
  * Copyright (C) 2004-2007 Cavium Networks
+ * Copyright (C) 2008 Wind River Systems
  */
 #include <linux/init.h>
 #include <linux/delay.h>
@@ -39,7 +40,6 @@ extern void pci_console_init(const char *arg);
 #ifdef CONFIG_CAVIUM_OCTEON_BOOTBUS_COMPACT_FLASH
 extern void ebt3000_cf_enable_dma(void);
 #endif
-static unsigned long CYCLES_PER_JIFFY;
 static unsigned long long MAX_MEMORY = 512ull << 20;
 
 
@@ -95,70 +95,12 @@ static void octeon_halt(void)
 
 
 /**
- * Read the Octeon high performance counter
- *
- * @return The counter value. For some brain dead reason, the kernel
- *         uses a 32bit number here.
- */
-static cycles_t octeon_hpt_read(void)
-{
-	cycles_t cycles;
-      asm("rdhwr %0,$31":"=r"(cycles));
-	return cycles;
-}
-
-
-/**
- * Acknowledge a timer tick. We don't use the standard Mips
- * one because it confuses the timer ticks and the HPT clock.
- */
-static void octeon_timer_ack(void)
-{
-	uint32_t count;
-	uint32_t next_compare = read_c0_compare() + CYCLES_PER_JIFFY;
-	write_c0_compare(next_compare);
-	count = read_c0_count();
-	if ((count - next_compare) < 0x7fffffff) {
-		next_compare = count + CYCLES_PER_JIFFY;
-		write_c0_compare(next_compare);
-	}
-}
-
-
-/**
- * Interrupt entry point for timer ticks
- *
- * @param irq
- * @param dev_id
- * @return
- */
-static irqreturn_t octeon_main_timer_interrupt(int irq, void *dev_id)
-{
-	if (read_c0_cause() & (1 << 30)) {
-		if (smp_processor_id() == 0) {
-			/* This function calls the timer ack internally */
-			timer_interrupt(irq, dev_id);
-		} else {
-			octeon_timer_ack();
-			local_timer_interrupt(irq, dev_id);
-		}
-		return IRQ_HANDLED;
-	} else
-		return IRQ_NONE;
-}
-
-
-/**
- * Setup the first cores timer interrupt
- *
- * @param irq
+ * Platform time init specifics.
  * @return
  */
-void __init plat_timer_setup(struct irqaction *irq)
+void __init plat_time_init(void)
 {
-	irq->handler = octeon_main_timer_interrupt;
-	irq->flags |= IRQF_SHARED;
-	setup_irq(7, irq);
+	/* Nothing special here, but we are required to have one */
 }
 
 
@@ -297,9 +239,6 @@ void prom_init(void)
 	}
 
 	mips_hpt_frequency = octeon_get_clock_rate();
-	clocksource_mips.read = octeon_hpt_read;
-	mips_timer_ack = octeon_timer_ack;
-	CYCLES_PER_JIFFY = ((mips_hpt_frequency + HZ / 2) / HZ);
 
 	_machine_restart = octeon_restart;
 	_machine_halt = octeon_halt;
-- 
1.5.5.1

